daml/ContingentClaims/Lifecycle.daml

```diff
@@ -3,83 +3,119 @@
--- 2. Function used when branching in an `Or`. Must return one of the arguments.
--- 3. A set of `Claim`s.
--- 4. A function which, given a date, lifecycles the third argument.
-lifecycle : (Eq a, CanAbort m)
+-- | Collect claims falling due into a list, and return the tree with those nodes pruned.
+-- `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
+-- The first argument is used to lookup the value of any `Observables`.
+-- Results in a function taking today's date, and returning the pruned tree + pending settlements.
+settle : (Eq a, CanAbort m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
   -> C a
   -> Date -> m (Result a)
-lifecycle spot choose
+settle spot
   = runKleisli
   . fmap (uncurry Result)
-  . flip evalStateT 1.0
   . runWriterT
-  . apoCataM (pruneZeros') (lifecycle' spot choose)
+  . apoCataM pruneZeros' acquireThenSettle -- : (Decimal, C a) -> WriterT [(Decimal, a)] (Kleisli m Date) (F a (Either (C a) (Decimal, C a))) )
+  . (1.0 : Decimal, )
+    where
+      acquireThenSettle =   fmap (fmap distE)  -- TODO: simplify
+                          . fmap join 
+                          . fmap (fmap sequence) 
+                          . fmap (fmap (fmap (settle' spot))) 
+                          . fmap (fmap sequence) 
+                          . fmap sequence 
+                          . fmap 
+                          . fmap (fmap (fmap embed ) ) 
+                          . fmap (fmap sequence) 
+                          . fmap lift 
+                          $ acquire' spot
+      distE : Either (C a) (F a x) -> F a (Either (C a) x)
+      distE (Left c) = Left <$> project c
+      distE (Right f) = Right <$> f
 
-type Seed m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
-
-lifecycle' : (Eq a, Monad m, CanAbort m)
+-- | Evaluate observables and skip branches for which predicates don't hold.
+-- This is useless on its own. Composed with other functions it adds laziness.
+acquire' : (Eq a, Monad m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
-  -> C a -> Seed m a
+  -> C a -> Kleisli m Date (F a (Either (C a) (C a)))
+acquire' spot (When obs c) = do
+  predicate <- eval spot obs
+  if predicate then acquire' spot c else pure $ WhenF obs (Left c)
+acquire' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then acquire' spot c else acquire' spot c'
+acquire' _ other = pure $ Right <$> project other
+
+-- | Collects all `One` leaves replacing them with `Zero`s (to avoid double-spend).
+-- Calculates the product of any `Scale` factors between each `One` and the root node, logging these to the writer.
+-- This function **ignores** conditions; it traverses the entire tree.
+settle' : CanAbort m 
+        => (Text -> Date -> m Decimal) 
+        -> (Decimal, C a) -> WriterT [(Decimal, a)] (Kleisli m Date) (F a (Decimal, C a))
```

Comment:
Had to get rid of the 'quantity' state monad and put the carrier in the input argument. The reason is that with `StateT`, the quantity was being accumulated _across_ branches. It's like we need an `Applicative` `StateT` in order to not accumulate across-branches. This is the next best thing. This fixes #18 .